### ISSUE TITLE
Clarify Accessible Authentication by including "remembering user names and passwords" in the SC text

### ISSUE THREAD

**Issue by WilcoFiers** [w3c/wcag#2577]
Success criterion 3.3.7 Accessible Authentication currently reads:

> For each step in an authentication process that relies on a [cognitive function test](https://w3c.github.io/wcag/guidelines/22/#dfn-cognitive-function-test)...

The difficulty with that is that "cognitive function test" is easily misunderstood to mean "CAPTCHA". It would be helpful to explicitly mention the most common case; remembering user names and passwords. The SC would read as follows with that addition:

> For each step in an authentication process that relies on a [cognitive function test](https://w3c.github.io/wcag/guidelines/22/#dfn-cognitive-function-test) **such as remembering a username and password**, at least one other authentication method is available that does not rely on a cognitive function test, or a mechanism is available to assist the user in completing the cognitive function test.

Alternatively:

> For each step in an authentication process that relies on **remembering a username and password, or some other** [cognitive function test](https://w3c.github.io/wcag/guidelines/22/#dfn-cognitive-function-test), at least one other authentication method is available that does not rely on a cognitive function test, or a mechanism is available to assist the user in completing the cognitive function test.

**Comment by dylanb** [comment]
This is a great suggestion. I think, in general, clarifying normative S.C. statements like you have suggested would improve the understandability of WCAG by the average software development team member.

**Comment by GreggVan** [comment]
+1 

That really changes the  it from abstract to concrete  and if username/password is considered a cognitive test by the WG is should certainly be plainly evident. 

Gregg Vanderheiden
***@***.***

**Comment by patrickhlauke** [comment]
i'd perhaps make it a parenthetical, or a normative note to surface this a bit more in the normative part (a la "Note: cognitive tests include remembering..."), but in principle yes i'm in favour of making it clearer that this is not just about CAPTCHAs

**Comment by GreggVan** [comment]
+1  
and if we think remembering a username and password is included — we really need to state that.   

gregg

———————————
Professor, University of Maryland, College Park
Founder and Director Emeritus , Trace R&D Center, UMD
Co-Founder Raising the Floor. http://raisingthefloor.org
The Global Public Inclusive Infrastructure (GPII) http://GPII.net
The Morphic project  https://morphic.org

**Comment by yatil** [comment]
Yes, please!

**Comment by bruce-usab** [comment]
I agree that username/password is a cognitive function test.  But that idea is already reinforced by the Note, so IMHO we do  need an edit.

Also, username/password is covered by the exception since those are <q>content the user provided to the website</q>.

**Comment by alastc** [comment]
Bruce, I think you might be thinking of an older version, the first bullet in CFT is:

> memorization, such as remembering a username, password, set of characters, images, or patterns. The common identifiers name, e-mail, and phone number are not considered cognitive function tests as they are personal to the user and consistent across websites;

**Comment by stevefaulkner** [comment]
+1 to this:
> For each step in an authentication process that relies on a [cognitive function test](https://w3c.github.io/wcag/guidelines/22/#dfn-cognitive-function-test) such as remembering a username and password, at least one other authentication method is available that does not rely on a cognitive function test, or a mechanism is available to assist the user in completing the cognitive function test.

**Comment by patrickhlauke** [comment]
> UPDATE: Sorry, I see what you mean about the exception: "Exception: When the cognitive function test is to recognize objects, or content the user provided to the website."
> 
> The intent behind 'content' is things like images, not words/passwords.

Let's make sure it's not misread this way, as it would otherwise make this SC rather pointless for one of the cases it tries to cover

**Comment by awkawk** [comment]
One problem with this suggested edit is that some usernames will be the user's email address, and I worry that this will be confusing to people. We could say that the username/password combination would be a problem (even if the username is the user's email because they may not remember that) and that email/password would only be problematic for the password input, but will implementers understand that?

**Comment by yatil** [comment]
I looked in the [Understanding document](https://w3c.github.io/wcag/understanding/accessible-authentication.html) – which might be outdated – and it explicitly states to allow usernames and passwords as long as they can be autofilled with a password manager or allow copy and paste:

> Examples
>
> - A web site uses a properly marked up username (or email) and password fields as the login authentication (meeting Success Criterion 1.3.5 Input Purpose and Success Criterion 4.1.2: Name, Role, Value). The user's browser or integrated third-party password manager extension can identify the purpose of the inputs and automatically fill in the username and password.
> - A web site does not block paste functionality. The user is able to use a third-party password manager to store credentials, copy them, and paste them directly into a login form.

I think the first bullet point is already covered with 1.3.5 and must be met anyway as both of them are Level AA, so I am unsure why this has any bearing on this SC as we would assume other AA SCs also need to be met.

Allowing these two techniques makes the whole SC not super useful, as the only change to be compliant with the SC is to either conform to 1.3.5 Input Purpose or allow copy and paste. If that is indeed the goal of this SC, then it should just be “Allow copy and paste into authentification form fields.” If the goal is to always have an alternative to username and password, the Understanding needs to reflect that.

**Comment by yatil** [comment]
And to the effect to this proposal, I think it needs to be explicit: 

> process that relies on remembering _and entering_ a username and password, or some other cognitive function test,

That said, I don’t feel like this is a strong SC in its current form.

**Comment by alastc** [comment]
>  I am unsure why this has any bearing on this SC as we would assume other AA SCs also need to be met.

Yes, there is overlap and following good practice will pass this SC in simple cases. However, one of the main problems that COGA brought to the group was 2nd factor auths which require you to type a number in.

Plus, you must have come across a username/password field which blocks password managers?

And, many sites are not being to be able to get rid of username/passwords soon, it wouldn't be feasible to require that.

Noting the username/email overlap above, how about:

> For each step in an authentication process that relies on a [cognitive function test](https://w3c.github.io/wcag/guidelines/22/#dfn-cognitive-function-test) **such as remembering a password**, at least one other authentication method is available that does not rely on a cognitive function test, or a mechanism is available to assist the user in completing the cognitive function test.

**Comment by GreggVan** [comment]
Good point Alastair

	- so the only failure around that would be if the site had password and username where it did NOT allow the use of password managers.  (Which by the way — I run into all the time)

I also note in another comment - that someone said that name role etc. are required elsewhere.  But you can pass that SC and not use the exactly correct name for example and password managers will still fail.   And there are other ways to pass that SC and still fail this one.   So I think this topic is still important to cover here. 


RE the proposed wording
> For each step in an authentication process that relies on a cognitive function test <https://w3c.github.io/wcag/guidelines/22/#dfn-cognitive-function-test> such as remembering a password, at least one other authentication method is available that does not rely on a cognitive function test, or a mechanism is available to assist the user in completing the cognitive function test.
> 

It think this is good, except that "assist" is vague and unclear.  What is considered enough of an assist?   Better directions? 
Also it sounds like the mechanism needs to be provided by the author. 

Thought it would be easy to fix - but — not really.  
Best I can think of is the following
Anyone else got a better idea? 

> For each step in an authentication process that relies on a cognitive function test <https://w3c.github.io/wcag/guidelines/22/#dfn-cognitive-function-test> such as remembering a username and password, at least one other authentication method is available that does not rely on a cognitive function test, unless the test can be passed using commonly available tools that do not require a cognitive function test to operate them.
> 



gregg

———————————
Professor, University of Maryland, College Park
Founder and Director Emeritus , Trace R&D Center, UMD
Co-Founder Raising the Floor. http://raisingthefloor.org
The Global Public Inclusive Infrastructure (GPII) http://GPII.net
The Morphic project  https://morphic.org

**Comment by giacomo-petri** [comment]
> unless the test can be passed using commonly available tools that do not require a cognitive function test to operate them.

also "commonly available tools" might be vague and unclear.
For example "copy and paste" commands, the autocomplete="one-time-code" attribute, etc. can be included into the definition of "commonly available tool"?

> or a mechanism is available to assist the user in completing the cognitive function test.

Maybe "assist" might be vague and unclear, but in combination with "in completing" assumes that the "test can be passed". Also "mechanism" in my opinion better fits with examples I've provided above.
Last, but not least, the Note about mechanisms

> Examples of mechanisms include: 1) support for password entry by password managers to address the memorization cognitive function test, and 2) copy and paste to help address the transcription cognitive function test.

helps understanding the intent, that can be expanded in understanding, if needed.

**Comment by yatil** [comment]
@alastc Good point, and I think just saying “such as remembering a password” is a good solution. Maybe even say:

> For each step in an authentication process that relies on a [cognitive function test](https://w3c.github.io/wcag/guidelines/22/#dfn-cognitive-function-test) such as **remembering a password or solving a puzzle**, at least one other authentication method is available that does not rely on a cognitive function test, or a mechanism is available to assist the user in completing the cognitive function test.

---

> Yes, there is overlap and following good practice will pass this SC in simple cases. However, one of the main problems that COGA brought to the group was 2nd factor auths which require you to type a number in.

Oh, right – because the 2FA code is not a personal attribute it is not covered by 1.3.5 Input Purpose.

> Plus, you must have come across a username/password field which blocks password managers?

Maybe a handful in the last 10-ish years of password manager uses.

> And, many sites will not be able to get rid of username/passwords anytime soon, it wouldn't be feasible to require that.

This is like saying “Many sites will not be able to add alternative texts to their images anytime soon, it wouldn’t be feasible to require that”. Let’s face it, it will be at least a year until 2.2 is widely adopted, maybe three until it is in legislation. Many companies push to password-less authentication at least as an option. I think 2.2 can raise the bar higher in this regard. (But I might be wrong.)

**Comment by alastc** [comment]
> This is like saying “Many sites will not be able to add alternative texts to their images anytime soon, it wouldn’t be feasible to require that”. 

I disagree with that pretty strongly. When managing a site, most of the CMSs and frameworks default to username/password. I don't think the password-less options are common yet. Partly because they are  new, and partly because they require external accounts/arrangements with providers like Apple/MS.

If you have 2-factor auth then you still need the default factor of a username/password anyway, it is just minimised in usage.

This SC was pushed back from 2.1 because WebAuthN wasn't mature/popular enough, things are changing, but slowly.

**Comment by bruce-usab** [comment]
> The intent behind 'content' is things like images, not words/passwords.

This is part of why it's important to write out the exception as a full complete sentence.  The suggested phrasing I made for that  might be a swing-and-a-miss then &mdash; because I did think "content" included ID/password provided by the end-user.

> That said, I don’t feel like this is a strong SC in its current form.

I think that is okay because the [3.3.8 No Exception](https://w3c.github.io/wcag/guidelines/22/#accessible-authentication-no-exception) (AAA) is the idealized version.
